src/bap_mc/bap_mc.ml

```diff
@@ -103,30 +110,41 @@ let print_bil lift bil_formats mem insn =
   List.iter bil_formats ~f:(fun fmt ->
       printf "%s@." (string_of_bil fmt (bil insn)))
 
-let make_print lift insn_fmt bil_fmt show_kinds mem insn =
-  print_insn insn_fmt insn;
+let make_print lift insn_fmt bil_fmt show_kinds oplen_bases mem insn =
+  print_insn_oplen oplen_bases mem;
+  print_insn insn_fmt insn; 
   print_bil lift bil_fmt mem insn;
   if show_kinds then print_kinds insn
 
-let step print state mem insn _ =
-  print mem insn;
+
+let check disasm_len current_total =
+  if disasm_len=0 then true
+  else current_total < disasm_len
+
+let step disasm_len hit_count print state mem insn _ =
+  if (check disasm_len !hit_count) then (
+    hit_count := !hit_count+1;
+    print mem insn;
+  ) else ();
```

Comment:
First of all there is no need to provide `else` part if `then` part evaluates to zero. Second, you have encoded quite misleading, behavior. You promised a user to disassemble only `N` instructions, but actually, you disassemble all the instructions, but print only first `N`. That is not nice, what if they gave you a string of length gigabyte? Or worse, what if after the first instruction, there is no valid instructions at all, but just data, as it is quite usual for `jmp` instructions in `x86`?
In order to stop disassembler, you need to call `Disasm.stop` instead of step. And yes, this means, that you need also update the logic of exhaustiveness check.